### PR TITLE
#290 Added optional S3 data store for BS CSV files

### PR DESCRIPTION
**Pull request by nickdos**
#290

### CODE REVIEW

File: grails-app/conf/application.yml [hunk 1]
```diff
@@ -219,6 +231,8 @@ environments:
         biosecurity:
             csv:
                 local:
+                    enabled: false
```
Comment (qifeng-bai) [inline grails-app/conf/application.yml:234]: I wonder if we should remove the local variable and set local as a default option.  If S3 is false,  we use local storage as default

Comment (nickdos) [reply grails-app/conf/application.yml:234]: Yes, good catch - missed this - should be local by default. I'll fix and push now.

Comment (nickdos) [reply grails-app/conf/application.yml]: Refresh to see changes.

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy [hunk 2]
```diff
@@ -21,25 +24,51 @@ import java.nio.file.StandardCopyOption
...
 class BiosecurityCSVService {
     def diffService
     def grailsApplication
+    AmazonS3Service amazonS3Service
 
-    def list() {
-        def BASE_DIRECTORY = grailsApplication.config.getProperty('biosecurity.csv.local.directory', '/tmp')
-        if (grailsApplication.config.getProperty('biosecurity.csv.local.enabled', Boolean, false)) {
+    def list() throws Exception {
+        if (grailsApplication.config.getProperty('biosecurity.csv.local.enabled', Boolean, true)) {
```
Comment (qifeng-bai) [inline grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:40]: Shall we move those codes to 'else' section?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy [hunk 3]
```diff
@@ -51,15 +80,19 @@ class BiosecurityCSVService {
...
             def records = diffService.getNewRecords(qs) // TODO: unused var - remove?
-
             File outputFile = createTempCSV(qs)
+            String folderName = new SimpleDateFormat("yyyy-MM-dd").format(new Date())
+            String fileName = sanitizeFileName("${qs.query.name}")+ ".csv"
 
-            if (grailsApplication.config.getProperty('biosecurity.csv.local.enabled', Boolean, false)) {
-                String fileName = sanitizeFileName("${qs.query.name}")+ ".csv"
-                String folderName = new SimpleDateFormat("yyyy-MM-dd").format(new Date())
+            if (grailsApplication.config.getProperty('biosecurity.csv.local.enabled', Boolean, true)) {
```
Comment (qifeng-bai) [inline grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:87]: Here as well?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy [hunk 4]
```diff
@@ -173,7 +206,27 @@ class BiosecurityCSVService {
         def BASE_DIRECTORY = grailsApplication.config.getProperty('biosecurity.csv.local.directory', '/tmp')
         def folder = new File(BASE_DIRECTORY, folderName)
         Collection<File> csvFiles = []
-        collectCsvFiles(folder, csvFiles)
+
+        if (grailsApplication.config.getProperty('biosecurity.csv.local.enabled', Boolean, true)) {
```
Comment (qifeng-bai) [inline grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:210]: Same?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy [hunk 5]
```diff
@@ -202,16 +255,175 @@ class BiosecurityCSVService {
...
+     *
+     * @param filename
+     * @return file content as String
+     */
+    String getFile(String filename) {
+        if (!filename) return ''
+
+        def config = grailsApplication.config
+
+        if (config.getProperty('biosecurity.csv.local.enabled', Boolean, true)) {
```
Comment (qifeng-bai) [inline grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:307]: Remove?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy [hunk 6]
```diff
@@ -202,16 +255,175 @@ class BiosecurityCSVService {
...
+
+        ''
+    }
+
+    Map deleteFile(String filename) {
+        if (!filename) return [:]
+
+        def config = grailsApplication.config
+
+        if (config.getProperty('biosecurity.csv.local.enabled', Boolean, true)) {
```
Comment (qifeng-bai) [inline grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:332]: Move to 'else' ?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy (hunk 2 above)
Comment (nickdos) [reply grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:40]: Don't understand comment. Do you mean swap local and S3 if order?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy (hunk 5 above)
Comment (nickdos) [reply grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:307]: Remove what? The `if (config.getProperty('biosecurity.csv.local.enabled', Boolean, true)) {` line? For what reason?

File: grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy (hunk 2 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/BiosecurityCSVService.groovy:40]: I think we should not check this variable after we remove 'biosecurity.csv.local' from config

**Comment by nickdos** [comment]
@bai - most of those comments have been dealt with with fixes. Anything else of note or good to approve?

**Review by qifeng-bai — APPROVED** [review]

**Pull request merged**